Initial support for Cmake

tete17 opened the pull request:
It may not be much but it is a start.

I am mostly familiar with best practices consuming libraries but when it comes to creating them my knoledge could be improve.

I will do some resaerch now but this is enoguh code so that others can take a look at it.

Feedback is welcomed.

#23 Is related

SuperWig commented:
`cxx_std_14` is CMake 3.8 isn't it? Yet you have 3.1 for the min.

liff-engineer commented:
If you want get full CMake support,make sure user can used like this:
```CMAKE
find_package(cli REQUIRED)

target_link_libraries(target PRIVATE cli::cli)
```

You should finish three step:
1. build 
2. install 
3. export

Here is example:
```CMAKE
cmake_minimum_required(VERSION 3.13)

project(cli 
    LANGUAGES CXX 
    VERSION 1.0
    )

############################
## build action: just build
############################

##Ensure boost::asio exist. try google "cmake boost asio" find why.
find_package(Boost REQUIRED COMPONENTS system)

add_library(cli INTERFACE)

##target alias for export
add_library(cli::cli ALIAS cli)

##require c++14
target_compile_features(cli INTERFACE cxx_std_14)

##config include dir for build and export use
target_include_directories(cli
        INTERFACE 
        $<INSTALL_INTERFACE:include>
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}> 
    )

target_link_libraries(cli INTERFACE Boost::system)

############################
## install action: install build result to specified location
############################

## install  include files 
install(DIRECTORY cli DESTINATION include)

## install target to cliTargets.cmake
## write to build dir
## cliTargets.cmake define IMPORTED library 
install(TARGETS cli
    EXPORT cliTargets
    RUNTIME DESTINATION bin
    ARCHIVE DESTINATION lib
    LIBRARY DESTINATION lib
    INCLUDES DESTINATION include    
)

############################
## export action: used for find_package support
############################

## install cliConfig.cmake to support find_package(cli)
## cliConfig.cmake should list required librarys and import cliTargets.cmake
file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/cliConfig.cmake
"include(CMakeFindDependencyMacro)

find_package(Boost REQUIRED COMPONENTS system)

if(NOT TARGET cli::cli)
    include(\"\${CMAKE_CURRENT_LIST_DIR}/cliTargets.cmake\")
endif()
"
)

## version config file also need
## write to build dir
include(CMakePackageConfigHelpers)
write_basic_package_version_file(
    cliConfigVersion.cmake 
    VERSION ${PACKAGE_VERSION}
    COMPATIBILITY AnyNewerVersion
)

## install config file to cmake script location
install(EXPORT cliTargets
    FILE cliTargets.cmake
    NAMESPACE cli::
    DESTINATION lib/cmake/cli)

install(FILES 
    ${CMAKE_CURRENT_BINARY_DIR}/cliConfig.cmake  
    ${CMAKE_CURRENT_BINARY_DIR}/cliConfigVersion.cmake  
    DESTINATION lib/cmake/cli)
```
Now build and install:

```BAT
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX:PATH=cli_install_location
cmake --build . --target install
```

And use it:

```CMAKE
##if you assign install path to  cli_install_location,you should config like this
list(APPEND CMAKE_PREFIX_PATH "cli_install_location")

find_package(cli REQUIRED)
target_link_libraries(target PRIVATE cli::cli)
```

Sorry for write it in this comment, I don't know how to send a PR.

liff-engineer commented:
Now vcpkg support will be easy:

CONTROL file is simple:
```TXT
Source:cli
Version:1.0
Description:A library for interactive command line interfaces in modern C++
Build-Depends:boost-asio
```
"portfile.cmake" like this:
```CMAKE
include(vcpkg_common_functions)

vcpkg_download_distfile(ARCHIVE
    URLS "https://github.com/daniele77/cli/archive/v1.0.0.tar.gz"
    FILENAME "cli-v1.0.0.tar.gz"
    SHA512  0000  #run "vcpkg install cli"  would show real SHA512,just replace
)

vcpkg_extract_source_archive_ex(
    OUT_SOURCE_PATH SOURCE_PATH
    ARCHIVE ${ARCHIVE}
)

vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
)

vcpkg_install_cmake()

file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/)

file(INSTALL ${SOURCE_PATH}/LICENSE DESTINATION ${CURRENT_PACKAGES_DIR}/share/cli/copyright)
```

daniele77 commented:
Great work, thanks! Just one request: would it be possible to keep the header files in the "/cli" directory instead of moving everything to "/include/cli"?

liff-engineer commented:
Use this:
```CMAKE
target_include_directories(cli
        INTERFACE 
        $<INSTALL_INTERFACE:include>
        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}> 
    )
```
BUILD_INTERFACE use to config build time include;just config to current source dir,don't need move header files.

In best pratices ,you should move header files to "include/library-name/".

tiendq commented on CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+find_package(Boost 1.55 REQUIRED COMPONENTS system)
+find_package(Threads REQUIRED)
+
+add_library(cli INTERFACE)
```
Should we use a variable for target name e.g. built-in`${PROJECT_NAME}`?

tete17 replied:
I think it could be a good idea since the name of the project may change in the future (I personally think cli is way too generic).

But i also think the ultimate call must be made by @daniele77

daniele77 replied:
@tete17 I completely agree. "cli" is definitely too generic but -- you know -- it was an internal project and the name dates back to the days when I wouldn't imagine it would be used by others.

So:

1. @tiendq Ok, let's use a variable name for the target name
2. Let's think to a good name. Any idea? Something like "clipp" but not so abused. If we don't come up with a good descriptive name, I'll go for something completely uncorrelated to the library.
3. Do you know which are the issues in changing the name of a github project? The link, the repository name...? What kind of problems I will be facing?

Thanks, guys!

tiendq commented on CMakeLists.txt:
```diff
@@ -2,12 +2,21 @@ cmake_minimum_required(VERSION 3.1)
+target_include_directories(cli
+        INTERFACE
+        $<INSTALL_INTERFACE:include>
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
```
There is no `include` directory.

daniele77 commented:
The Cmake version required should not depend on the c++ standard, but only on the cmake features are used in the CMakeLists.txt script.
E.g., on my old Debian box, I have cmake version 3.6.2 with gcc v. 4.9.2 (C++14). Besides, on my machine, I can have an old cmake but the latest gcc version that I compiled from sources.

SuperWig commented:
This line right here is what is requiring CMake 3.8
```cmake
target_compile_features(cli INTERFACE cxx_std_14)
```
`cxx_std_14` is a CMake 3.8 feature [3.7](https://cmake.org/cmake/help/v3.7/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html) vs [3.8](https://cmake.org/cmake/help/v3.7/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)

daniele77 commented:
> This line right here is what is requiring CMake 3.8
> ```cmake
> target_compile_features(cli INTERFACE cxx_std_14)
> ```
> `cxx_std_14` is a CMake 3.8 feature [3.7](https://cmake.org/cmake/help/v3.7/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html) vs [3.8](https://cmake.org/cmake/help/v3.8/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)

Well, actually on a old debian box of mine, with cmake v. 3.6 installed, the cmake script works if I change the first line to:

 ```cmake_minimum_required(VERSION 3.6)```

So, I guess cmake does not perform checks on features it does not know about?

tete17 commented:
Sorry for disappearing for a couple of days.
Work is a bit demanding.

@liff-engineer I am afraid I don't feel comfortable enough to create the vcpkg package since I am a Linux user, bur hopefully after this Cmake support it would be use for you.

Regarding the include directories, It is true not having a include directory may not be the best of practices since you are including the main directory of the project and you are allowed to include (during development) that you should not since they may not be installed on your users directory. If you want to keep the structure of the project we could also include the cli project which would mean your end users would use your library in this way instead
```c++
#include <cli.h>
```

Regarding cmake versions, cmake deals with version in a really extraneous way that after using it for more than 2 years professionally still evades my understanding :laughing: :laughing: :laughing: 
The rest of cmake is really cool though

tete17 commented:
> This line right here is what is requiring CMake 3.8
> 
> ```cmake
> target_compile_features(cli INTERFACE cxx_std_14)
> ```
> 
> `cxx_std_14` is a CMake 3.8 feature [3.7](https://cmake.org/cmake/help/v3.7/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html) vs [3.8](https://cmake.org/cmake/help/v3.8/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html)
> 
> I guess you could make it 3.1 and `cxx_generic_lambdas`?
> 
> Should probably add CMake files to the gitignore?

@SuperWig 
What files do you mean by CMake ones?? 
The CMakeLists.txt file must be tracked by git is part of the project

SuperWig commented:
Well obviously I don't mean the CMakeLists.txt as that's what this PR is adding. I'm talking about the CMake generated files. 

CMakeFiles/ 
CMakeCache.txt
install_manifest.txt
etc.

https://github.com/github/gitignore/blob/master/CMake.gitignore

tete17 commented:
I will add them to the `.gitignore` but cmake usually expects you to do out of source build typically on a `build` directory or in the case of Clion `cmake-build-*` folders.

SuperWig commented:
@daniele77 This script wouldn't be using it. It would be the projects that then link to cli::cli and you definitely get an error with CMake 3.6

@tete17 Out of source doesn't necessarily mean out of repo.

---

I'm no CMake expert (or beginner really) but every packageConfig.cmake I've seen look more like

```cmake
include(CMakeFindDependencyMacro)

find_dependency(Boost REQUIRED COMPONENTS system)

if(NOT TARGET cli::cli)
    include(${CMAKE_CURRENT_LIST_DIR}/cliTargets.cmake)
endif()
```
With `find_dependency` instead of package. @liff-engineer did you mean to do that before? Because you included it but didn't use it

ras0219-msft commented:
>I am afraid I don't feel comfortable enough to create the vcpkg package since I am a Linux user

FWIW @tete17, vcpkg can be used on linux :)

tete17 commented:
One quick question @SuperWig what do you mean by `Out of source doesn't necessarily mean out of repo.`

I afraid I don't follow :thinking: 

##

@ras0219-msft Hope you can forgive my ignorance. I will take a look at the project in the future

SuperWig commented:
Making a build directory within the repository directory.

daniele77 commented:
After thinking about it I agree with the original idea:

- have sources in the directory `include/cli` (it seems it's the standard convention)
- use `#include <cli/cli.h>` in the sources

(sorry for the unnecessary work I asked to do).

I have a couple of additional requests:

- I'd like to keep my old `samples/Makefile`, `samples/makefile.win` and `samples/samples.sln` files. I know that with cmake they're unnecessary, but for the moment I'd prefer to keep them for those who doesn't use cmake. I think these files don't bother because you usually launch cmake in a subdirectory (e.g., `build`).
- Can you add a couple of lines in the `README` file to explain how to compile the examples with cmake? You know: the usual exporting `BOOST_INCLUDEDIR`, create `build` directory, `cmake ..` and `make`

Finally: someone suggested me to change the name of the directory "samples" to "examples", because it's semantically more correct. What do you think about it?

Thanks a lot for contributing positively.

SuperWig commented:
So as per the comments made https://github.com/daniele77/cli/pull/34#issuecomment-468941998 here then I figured the samples/examples CMakeLists.txt should be something like this?
```cmake
set(SOURCES complete filesession simplelocalsession)
if(NOT WIN32)
  list(APPEND SOURCES asyncsession)
#include message?
else()
  message("asyncsession not built as it's not supported on Windows")
endif()

foreach(example ${SOURCES})
  add_executable(${example} ${example}.cpp)    
  target_link_libraries(${example} cli::cli)
endforeach(example)
```

daniele77 commented:
> So as per the comments made #34 (comment) here then I figured the samples/examples CMakeLists.txt should be something like this?

@SuperWig exactly

daniele77 commented:
Please, could someone explain to me step-by-step how to compile the examples using cmake? I tried but I had no success :-(

SuperWig commented:
```
mkdir build && cd build
cmake .. -DCLI_BuildSamples=ON
cmake --build .
```

tete17 commented:
Sorry for not giving any life signs in a while I will update the PR today

daniele77 commented:
@SuperWig ok, it works. Thank you a lot.

@tete17 Great: I'm just starting to merge your contribution into the project.

tete17 commented:
I rebased and squashed the commits for a nice git history. 
Hopefully I have not forgotten anything, I so please let me know.
I am not sure what the problem with `asyncsession.cpp` in Windows platforms but I trust @SuperWig

daniele77 commented:
Thank you guys!
I merged your work in the master branch.
I had to make some little changes because there was a little bug in the installation path.

Can someone add vcpkg support, now? :-)

SuperWig commented:
Can someone add vcpkg support, now? :-)

Made the pull request: https://github.com/Microsoft/vcpkg/pull/5762

daniele77 commented:
@SuperWig I just released v. 1.1 : how can I tell vcpkg there is a new version of the library?

SuperWig commented:
clone vcpkg, then go to your port `ports/cli`
update `CONTROL` with the new version
update `portfile.cmake` REF to v1.1.0 and the SHA512 to whatever that is, attempting to install cli should give you the new hash.

vcpkg has commands to make this easier

```
./vcpkg edit cli
if you already have the tar downloaded
./vcpkg hash path/to/cli.tar
```

Then commit and  make a pull request.

daniele77 commented:
Great.
Have you an idea why vcpkg builds and installs all the boost libraries (everything! the math libraries too!) even if the cli CONTROL file depends only from boost-asio?

SuperWig commented:
Because boost-asio depends on boost.
https://github.com/Microsoft/vcpkg/blob/master/ports/boost-asio/CONTROL

daniele77 commented:
sigh :-(

ras0219-msft commented:
More specifically, boost-asio uses boost-date-time which uses boost-math.

daniele77 commented:
uhm... according to [this page](http://think-async.com/Asio/boost_asio_1_3_1/doc/html/boost_asio/using.html#boost_asio.using.dependencies
) the only mandatory dependency is Boost.System.
And if I remember well from having used it several times, Boost.Date-time does not depend on Boost.Math.
